src/upnp.cpp

```diff
@@ -708,6 +708,22 @@ void upnp::post(upnp::rootdevice const& d, char const* soap
 #endif
 }
 
+std::string upnp::create_soap(char const* soap_action, std::string const& service_namespace, char const* part)
```

Comment:
also, I imagine one benefit of factoring this out into its own function is that it can more easily be tested directly.
and, should this be a member function? (or free function?)